File: src/main/java/akka/actor/MasterActor.java
package akka.actor;

import akka.message.Result;


/**
 * @user sun
 * @date 2016/7/4
 */
public class MasterActor extends UntypedActor {

    private ActorRef aggregateActor = getContext().actorOf(Props.create(AggregateActor.class), "aggregate");

    private ActorRef reduceActor = getContext().actorOf(Props.create(
            (UntypedActorFactory) () -> new ReduceActor(aggregateActor)), "reduce");

    private ActorRef mapActor = getContext().actorOf(Props.create(
            (UntypedActorFactory) () -> new MapActor(reduceActor)), "map");

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof String) {
            mapActor.tell(message, getSelf());
        } else if (message instanceof Result) {
            aggregateActor.tell(message, getSelf());
        } else
            unhandled(message);
    }
}

File: src/main/scala/scala_akka/actor/MasterActor.scala
package scala_akka.actor

import akka.actor.{Actor, ActorRef, Props}

import scala_akka.message.Result

/**
  * @user sun 
  * @date 2016/7/10
  */
class MasterActor extends Actor {
  val aggregateActor: ActorRef = context.actorOf(Props[AggregateActor], "aggregate")
  val reduceActor: ActorRef = context.actorOf(Props(new ReduceActor(aggregateActor)), "reduce")
  val mapActor: ActorRef = context.actorOf(Props(new MapActor(reduceActor)), "map")

  override def receive: Receive = {
    case message: String => mapActor ! message
    case message: Result => aggregateActor ! message
    case _ =>
  }
}
